### PR TITLE
Fix cpp generator

### PR DESCRIPTION
**Pull request by BjoernAtBosch**
## Describe your changes

Fix C++ generator using matching namespace/folder structure

## Issue ticket number and link

AB#12882

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.

### CODE REVIEW

**Comment by BjoernAtBosch** [comment]
@doosuu:
I would go here for version 0.3.0 because of incompatible change in model generation.

File: sdv/model_generator/cpp/cpp_generator.py [hunk 1]
```diff
@@ -17,89 +17,105 @@
...
             root (_type_): the vspec tree root node.
         """
-        self.root = root
+        self.root_node = root_node
         self.target_folder = target_folder
         self.ctx_header = CodeGeneratorContext()
         self.includes: Set[str] = set()
         self.external_includes: Set[str] = set()
-        self.root_namespace = namespace
+        if "::" in root_namespace:
```
Comment (doosuu) [inline sdv/model_generator/cpp/cpp_generator.py]: For reduction of complexity: Introduce `__get_root_namespace_list(namespace: str) -> list[str]`?

File: sdv/model_generator/cpp/cpp_generator.py [hunk 2]
```diff
@@ -257,11 +275,13 @@ def __gen_nested_class(
...
         return class_code_context.get_content()
 
-    def __gen_collection_types(self, node: VSSNode, path: str) -> str:
+    def __gen_collection_types(self, node: VSSNode, namespace_list: List[str]) -> str:
         collection_types = []
         for child in node.children:
             if child.type == VSSType.BRANCH:
-                self.includes.add(f"{path}/{child.name}/{child.name}")
+                child_namespace_list = namespace_list + [child.name]
+                path = os.path.join("", *self.__to_folder_names(child_namespace_list))
```
Comment (doosuu) [inline sdv/model_generator/cpp/cpp_generator.py]: You can simply call `os.path.join(*self.__to_folder_names(child_namespace_list))` the empty string as first argument is superfluous

File: sdv/model_generator/cpp/cpp_generator.py [hunk 3]
```diff
@@ -257,11 +275,13 @@ def __gen_nested_class(
...
+                self.includes.add(f"{path}/{child.name}")
```
Comment (doosuu) [inline sdv/model_generator/cpp/cpp_generator.py]: Above you use `os.path.join` to do path based joining, here you rely on the literal `/` I would recommend to append `child_name` to the list before calling `os.path.join` to get the full path via join

File: sdv/model_generator/utils.py [hunk 4]
```diff
@@ -12,10 +12,18 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
-"""CodeGeneratorContext."""
+import re
 from typing import Optional
 
 
+def to_snake_case(camel_case_input):
```
Comment (doosuu) [inline sdv/model_generator/utils.py]: I expect that this only works for converting camel case to snake case. Maybe `camel_to_snake_case(input: str) -> str` would be a better way of calling this? Also please add type annotations

**Review by doosuu — CHANGES_REQUESTED** [review]

File: sdv/model_generator/cpp/cpp_generator.py (hunk 2 above)
Comment (BjoernAtBosch) [reply sdv/model_generator/cpp/cpp_generator.py]: Thought, I tried that and was not working. But yes, works. 
Thx!

**Review by doosuu — APPROVED** [review]
LGTM

**Pull request merged**